File: Java/Bible.java
import java.util.ArrayList;

public class Bible {
  // Width of window, scale,
  public final int width = 1225 * 10;
  public final int scale = 10;
  // Arc scale factor; 1 makes the final vis a circle
  public final double arc_height = 1;
  // Offset (for empty space on sides)
  public final int offset = (width - (1189 * (int)scale)) / 2;
  // List of all cross references
  public ArrayList<Arc> arcs;
  // List of first chapters in books
  public ArrayList<Chapter> chapters;

  // Constructor
  public Bible(){
    arcs = new ArrayList<Arc>();
    chapters = new ArrayList<Chapter>();
  }

  // Add a cross reference to list of cross references from a String input
  public void addArc(String line) {
    // Split line by comma
    String[] split_line = line.split(",");
    // Get data for Arc object
    int from_chapter_bible = Integer.parseInt(split_line[3]);
    int from_book_number = Integer.parseInt(split_line[1]);
    int to_chapter_bible = Integer.parseInt(split_line[8]);
    int to_book_number = Integer.parseInt(split_line[6]);
    // Create Arc object and add to list
    arcs.add(new Arc(from_chapter_bible, from_book_number, to_chapter_bible, to_book_number));
  }

  // Add a cross reference to list of cross references from a String input
  public void addChapter(String line) {
    // Split line by comma
    String[] split_line = line.split(",");
    // Get data for Chapter object
    int chapter_in_bible = Integer.parseInt(split_line[0]);
    String book_name = split_line[1];
    int chapter_in_book = Integer.parseInt(split_line[3]);
    // Save chapter if it's the first in its book
    if (chapter_in_book == 1) {
      chapters.add(new Chapter(chapter_in_bible, book_name));
    }
  }
}

File: C++/Bible.cpp
#include "Bible.hpp"
#include "Chapter.hpp"
#include "Arc.hpp"

#include <iostream>


// Constructor
Bible::Bible() {
	// Initialize vector to hold cross reference Arc objects
	refs = new std::vector<Arc>;
}

// Default destructor
Bible::~Bible() {
}

// Function to make Arc from a string input
Arc Bible::MakeArcFromLine(std::string line) {
	// Index of comma in input, declare vector of strings to hold data
        std::size_t comma = line.find(",");
        std::vector<std::string> vals;
	// While comma index is not out of bounds
        while (comma != std::string::npos) {
		// Add substring from 0 to comma to vals vector
                vals.push_back(line.substr(0, comma));
		// Update line to start after comma
                line = line.substr(comma + 1);
		// Find index of next comma
                comma = line.find(",");
        }
	// Construct from and to Chapter objects from vals vector
	Chapter* from = new Chapter(std::stoi(vals.at(3)), vals.at(0), std::stoi(vals.at(1)), std::stoi(vals.at(2)));
	Chapter* to = new Chapter(std::stoi(vals.at(8)), vals.at(5), std::stoi(vals.at(6)), std::stoi(vals.at(7)));
	// Return Arc object with the two Chapter objects
	return Arc{from, to};
}

// Add an Arc object to Bible given a string input
void Bible::AddArc(std::string line) {
	// Call MakeArcFromLine on input and add to refs vector
	refs->push_back(MakeArcFromLine(line));
}


// Render Bible on given window
void Bible::Render(sf::RenderWindow& window) {
	// Draw all cross reference Arc objects
	for (int i = 0; i < refs->size(); i++) {
		Arc a = refs->at(i);
		a.DrawArc(window);
	}
	
	// Draw horizontal white line to represent x-axis of all chapters
	sf::RectangleShape line(sf::Vector2f(refs->at(0).width - refs->at(0).offset * 2, 2));
	line.setPosition(sf::Vector2f(refs->at(0).offset, refs->at(0).height/2 - 1));
	
	window.draw(line);
}


